### ISSUE TITLE
Unexpected result from BinaryCrossEntropy loss calculation in keras fit/evaluate

### ISSUE THREAD

**Issue by will-cern** [keras-team/keras#17125]
Hello,

I am opening this issue here in response to the potential bug described here: https://discuss.tensorflow.org/t/how-is-binarycrossentropy-loss-calculated-in-keras-fit-evaluate/11331/11  The reproducer code is as follows (in tf 2.8.2):

```
from tensorflow import keras
import numpy as np
loss_func = keras.losses.BinaryCrossentropy()
nn = keras.Sequential([
  keras.layers.Dense(2**8, input_shape=(1,), activation='relu'),
  keras.layers.Dense(2,activation='softmax')
])
nn.compile(loss=loss_func,optimizer='adam')
train_x = np.array([0.4]) # this is an arbitrary input
train_y = np.array([[1.,0.]])
train_q = nn.predict(train_x)
print("train_q = ",train_q)
print("Evaluated loss = ",nn.evaluate(train_x,train_y))
print("Function loss = ",loss_func(train_y,train_q).numpy())
print("Manual loss = ", -np.log(train_q[0,0]) )
```

The last two ways of calculating the loss agree and are what I would expect, but the value returned by the `evaluate` method is different and unexplainable for me. 

Any help in debugging this issue would be much appreciated.

Thanks
Will

**Comment by PierrickPochelu** [comment]
I don't really understand why and how evaluate() returns this loss value. However, if you ask the "crossentropy" metrics you will have matching results.

```
from tensorflow import keras
import numpy as np

loss_func = keras.losses.BinaryCrossentropy()
nn = keras.Sequential([
  keras.layers.Dense(2**8, input_shape=(1,), activation='relu'),
  keras.layers.Dense(2,activation='softmax')
])
nn.compile(loss=loss_func,metrics=['crossentropy'],optimizer='adam')
train_x = np.array([0.4]) # this is an arbitrary input
train_y = np.array([[1.,0.]])
train_q = nn.predict(train_x)
print("train_q = ",train_q)
def CE(target,output,epsilon=1e-7):
    bce = target * np.log(output + epsilon)
    return -np.sum(bce)
print("Evaluated loss = ",nn.evaluate(x=train_x,y=train_y)) #first number is the loss second is the metrics
print("Function loss = ",loss_func(train_y,train_q).numpy())
print("Manual CE = ", CE(train_y,train_q))

```

returns:


train_q =  [[0.49278286 0.50721717]]
1/1 [==============================] - 1s 587ms/step - loss: 0.7004 - crossentropy: 0.7077
Evaluated loss =  [0.7004469633102417, **0.707686**6030693054]
Function loss =  **0.707686**54
Manual CE =  **0.707686**4838600159


This is the expected precision with FP32.

**Comment by sushreebarsa** [comment]
@will-cern Could you refer the comment above and keep us informed if it helps?
Thank you!

**Comment by will-cern** [comment]
So I can comment that I had also observed no issues whatsoever with categorical crossentropy loss function evaluation. I am not sure what the "crossentropy" option would correspond to, but if it corresponds to categorical crossentropy then that would be consistent.

My problem has always been restricted to the binary crossentropy loss function applied to networks with more than one output. This discovery came about from working with one of my students who had set up a network with two outputs but was training it on binary cross entropy. They either should have made the network output one value (the logit) or use the categorical crossentropy loss. Both of these options seemed to work in a way I understood but it was still completely puzzling to me what keras/tf is calculating in the case of binary crossentropy with more than one output. I'm not saying I would want to do this, but I really wanted to understand *what* it was doing since keras allows you to do this, but the result makes no sense???

**Comment by PierrickPochelu** [comment]
I dive deeper into BinaryCrossentropy, here is 2 important pieces of information:

1)
There are 2 potential cross entropy distance:  Penalize when label=1 and pred=0 --> `loss=label * log(pred)`. The other one penalizes also when label=0 and pred=1 --> `loss=label*log(pred) + (1-label)*log(1-pred)` . It seems Binary_cross entropy is the second one.

2) 
You can apply cross-entropy on those cases:
- A) multi-class mono-label or "mutually exclusive classes" --> 1 Softmax layer
- B) multi-class n-labels --> n Softmax layers
- C) 2 classes n-labels --> 1 Sigmoid layer of size n . Binary cross entropy seems designed for this case.


Here is the new code:

```
from tensorflow import keras
import numpy as np

loss_func = keras.losses.BinaryCrossentropy()
nn = keras.Sequential([
  keras.layers.Dense(2**8, input_shape=(1,), activation='relu'),
  keras.layers.Dense(2,activation='sigmoid')
])
nn.compile(loss=loss_func,metrics=['binary_crossentropy'],optimizer='adam')
train_x = np.array([0.4]) # this is an arbitrary input
train_y = np.array([[1.,0.]])
train_q = nn.predict(train_x)
print("train_q = ",train_q)
def CE(target,output,epsilon=1e-7):
    bce = target * np.log( output +epsilon) + (1-target) * np.log(  1-output+epsilon)
    return -np.mean(bce)
print("Evaluated loss = ",nn.evaluate(x=train_x,y=train_y))
print("Function loss = ",loss_func(train_y,train_q).numpy())
print("Manual CE = ", CE(train_y,train_q))
```

Produces:

_train_q =  [[0.5013947 0.5099719]]
1/1 [==============================] - 0s 210ms/step - loss: 0.7018 - binary_crossentropy: 0.7018
Evaluated loss =  [**0.70182**71684646606, **0.70182**71684646606]
Function loss =  **0.70182**693
Manual CE =  **0.70182**69300460815

I hope it helps.

**Comment by PierrickPochelu** [comment]
The neural network output have different meaning according the final activation function (sigmoid or argmax). Therefore, the loss (or the person developing the loss) has different input expectations.

Let's take an example to illustrate:

We develop a model predicting if an animal is a mammal, and if it is oviparous. At first glance, let's consider they are 2 mutually exclusive classes. Those options are correct:
- OUTPUT_SHAPE: 1 scalar ACTIVATION: sigmoid to predict P(mammal) LOSS: trained with binary_crossentropy loss
- OUTPUT_SHAPE: 1 vector of 2 elements ACTIVATION: argmax to predict [P(mammal),P(oviparous)] or equivalently  [P(mammal),1-P(mammal)] , trained LOSS: categorical_crossentropy loss . Notice: The output layer has 2 times more trainable parameters. So comparison, is not  100% fair between different losses/activations.


Now we discover that the platypus is a mammal which is also oviparous. Therefore, we discover that P(mammal) and P(oviparous) or not mutually exclusive anymore.

Those options are correct:
- OUTPUT SHAPE: 1 vector of 2 elements ACTIVATION: sigmoid to predict [P(mammal),P(egg)] LOSS: binary_crossentropy
- OUTPUT_SHAPE: 2 vectors (e.i, 2 different layers side-by-side) of 2 elements ACTIVATION: 2 argmax to predict [P(mammal),1-P(mammal)] and P(oviparous),1-P(oviparous)] LOSS: it produces 2 losses, trained with categorical_crossentropy. This solution is less elegant but correct.

**Comment by sampathweb** [comment]
@will-cern - Apologize for the delay in responding to this issue.  I get the same answer for all three functions in the code shared by @PierrickPochelu in https://github.com/keras-team/keras/issues/17125#issuecomment-1273185297.  You can also set `keras.utils.set_random_seed(0)` to get reproducible results in this example.

Is this still an open issue for you?

**Comment by sampathweb** [comment]
Going to close this issue. Please re-open if the above comment doesn't resolve it for you.
